tests/base/src/test/java/org/wildfly/security/ssl/SSLv2HelloAuthenticationTest.java

```diff
@@ -0,0 +1,348 @@
+                .build();
+
+        List<Protocol> list = new ArrayList<>();
+        list.add(Protocol.forName("SSLv2Hello"));
+        list.add(Protocol.forName("TLSv1"));
+
+        SSLContext serverContext = new SSLContextBuilder()
+                .setSecurityDomain(securityDomain)
+                .setKeyManager(getKeyManager("/ca/jks/scarab.keystore"))
+                .setTrustManager(getCATrustManager())
+                .setNeedClientAuth(true)
+                .setProtocolSelector(ProtocolSelector.empty().add(EnumSet.copyOf(list)))
+                .build().create();
+
+        SecurityIdentity identity = performConnectionTest(serverContext, "protocol://test-two-way-sslv2hello.org", "wildfly-ssl-test-config-v1_6.xml");
+        assertNotNull(identity);
+        assertEquals("Principal Name", "ladybird", identity.getPrincipal().getName());
+    }
+
+    private SecurityIdentity performConnectionTest(SSLContext serverContext, String clientUri, String clientConfigFileName) throws Exception {
+        System.setProperty("wildfly.config.url", SSLAuthenticationTest.class.getResource(clientConfigFileName).toExternalForm());
+        AccessController.doPrivileged((PrivilegedAction<Integer>) () -> Security.insertProviderAt(new WildFlyElytronProvider(), 1));
+
+        AuthenticationContext context = AuthenticationContext.getContextManager().get();
+        AuthenticationContextConfigurationClient contextConfigurationClient = AccessController.doPrivileged(AuthenticationContextConfigurationClient.ACTION);
+        SSLContext clientContext = contextConfigurationClient.getSSLContext(URI.create(clientUri), context);
+
+        SSLServerSocketFactory sslServerSocketFactory = serverContext.getServerSocketFactory();
+
+        SSLServerSocket sslServerSocket = (SSLServerSocket) sslServerSocketFactory.createServerSocket(1111, 10, InetAddress.getLoopbackAddress());
+
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+        Future<SSLSocket> socketFuture = executorService.submit(() -> {
+            try {
+                System.out.println("About to connect client");
+                SSLSocket sslSocket = (SSLSocket) clientContext.getSocketFactory().createSocket(InetAddress.getLoopbackAddress(), 1111);
+                sslSocket.getSession();
+
+                return sslSocket;
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            } finally {
+                System.out.println("Client connected");
+            }
+        });
+
+        SSLSocket serverSocket = (SSLSocket) sslServerSocket.accept();
+        SSLSession serverSession = serverSocket.getSession();
+        SSLSocket clientSocket = socketFuture.get();
+        SSLSession clientSession = clientSocket.getSession();
+
+        try {
+            // The negotiated protocol must be TLSv1, as SSLv2Hello is only used to negotiate a safe protocol
+            assertEquals("TLSv1", serverSession.getProtocol());
```

Comment:
We should also assert that `serverSocket.getEnabledProtocols()` and `clientSocket.getEnabledProtocols()` return the two protocols that are configured.

Reply:
added